source/crud/tests/README.md

```diff
@@ -201,3 +201,315 @@ CommandSucceededEvents. Then, insert an invalid document (e.g. `{x: 1}`) and ass
+CommandStartedEvents. Perform a `hello` command using `client` and record the `maxWriteBatchSize` value contained in the
+response.
+
+Construct a `MongoCollection` (referred to as `collection`) with the namespace "db.coll" (referred to as `namespace`).
+Drop `collection`. Then, construct the following document (referred to as `document`):
+
+```json
+{
+  "_id": 1
+}
+```
+
+Insert `document` into `collection`.
+
+Create the following write model (referred to as `model`):
+
+```json
+InsertOne {
+  "namespace": namespace,
+  "document": document
+}
+```
+
+Construct a list of write models (referred to as `models`) with `model` repeated `maxWriteBatchSize + 1` times.
+
+#### Unordered
+
+Test that an unordered bulk write collects `WriteError`s across batches.
+
+Execute `bulkWrite` on `client` with `models` and `ordered` set to false. Assert that the bulk write fails and returns
+a `BulkWriteError` (referred to as `unorderedError`).
+
+Assert that `unorderedError.writeErrors` has a length of `maxWriteBatchSize + 1`.
+
+Assert that two CommandStartedEvents were observed for the `bulkWrite` command.
+
+#### Ordered
+
+Test that an ordered bulk write does not execute further batches when a `WriteError` occurs.
+
+Execute `bulkWrite` on `client` with `models` and `ordered` set to true. Assert that the bulk write fails and returns
+a `BulkWriteError` (referred to as `orderedError`).
+
+Assert that `orderedError.writeErrors` has a length of 1.
+
+Assert that one CommandStartedEvent was observed for the `bulkWrite` command.
+
+### 7. `MongoClient.bulkWrite` handles a cursor requiring a `getMore`
+
+Test that `MongoClient.bulkWrite` properly iterates the results cursor when `getMore` is required. This test creates
+a list of operation results that requires cursor iteration by performing inserts that each yield a `DuplicateKeyError`
+containing a very large `_id` value.
+
+This test must only be run on 8.0+ servers.
+
+Construct a `MongoClient` (referred to as `client`) with
+[command monitoring](../../command-logging-and-monitoring/command-logging-and-monitoring.rst) enabled to observe
+CommandStartedEvents. Perform a `hello` command using `client` and record the `maxBsonObjectSize` value from the
+response. Then, construct the following document (referred to as `document`):
+
+```json
+{
+  "_id": "a".repeat(maxBsonObjectSize / 2)
+}
+```
+
+Construct a `MongoCollection` (referred to as `collection`) with the namespace "db.coll" (referred to as `namespace`).
+Drop `collection`. Insert `document` into `collection`.
+
+Create the following write model (referred to as `model`):
+
+```json
+InsertOne {
+  "namespace": namespace,
+  "document": document
+}
+```
+
+Construct a list of write models (referred to as `models`) with `model` repeated 2 times. Execute `bulkWrite` on
+`client` with `models` and `ordered` set to false. Assert that the bulk write fails and returns a `BulkWriteError`
```

Comment:
Suggest also testing cursor iteration in a transaction: https://github.com/mongodb/specifications/commit/96656df8da7cc742b6d2ac8600faef1882c4fc4b

This may ensure drivers use the same session for the `bulkWrite` and `getMore`.

In the C implementation, not setting the same session on `getMore` results in an error:
```
Cannot run getMore on cursor 1983114598982261992, which was created in session 7ffcebae-ac99-40b4-b59f-830748d7d19f - 47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU= -  - , in session 9bf2e0e8-c903-460b-9f35-258c76777dbb - 47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=
```

Reply:
Added a test. The new test uses large upserted IDs rather than duplicate key errors to create a cursor with a nonzero ID because ordered:false is not respected within a transaction. I updated the other tests to use this strategy for consistency.